stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,407 @@
+   * ready.
+   * <br>
+   * If write is not legal at the time of call, immediately returns false
+   * <br><br>
+   * <b>NOTE:  </b>This method will return as soon as it passes the request to the grpc stream
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @return true if the request is sent to stream, false if skipped
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request) throws InterruptedException {
+    try {
+      return write(request, Integer.MAX_VALUE, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      throw new RuntimeException(e); // should never happen
+    }
+  }
+
+  /**
+   * Send a value to the stream for sending to server, wait if necessary for the grpc stream to be
+   * ready up to specified timeout.
+   * <br>
+   * If write is not legal at the time of call, immediately returns false
+   * <br><br>
+   * <p>
+   * <b>NOTE:  </b>This method will return as soon as it passes the request to the grpc stream
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * </p>
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @param timeout How long to wait before giving up.  Values &lt;= 0 are no wait
+   * @param unit A TimeUnit determining how to interpret the timeout parameter
+   * @return true if the request is sent to stream, false if skipped
+   * @throws TimeoutException if write does not become ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request, long timeout, TimeUnit unit)
+      throws InterruptedException, TimeoutException {
+
+    if (writeClosed) {
+      throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+    }
+
+    boolean writeDone = false;
+    long start = System.nanoTime();
+    long duration = unit.toNanos(timeout);
+
+    while (!writeDone && isWriteLegal()) {
+      if (call.isReady()) {
+        call.sendMessage(request);
+        writeDone = true;
+      } else {
+        waitAndDrainExecutorOrTimeout(start, duration);
```

Comment:
Passing start and duration is a bit strange. I'd have expected only passing the end time `start + duration` (inside it'd do `while (end - now() < 0)`) or the timeout `start + duration - now()` (and mostly delegate to Java APIs).

Reply:
LockSupport.parkNanos takes a number of nanos to wait rather than an end time so it fell out that way.  It could just as well be start and end time passed into the utility method as the current start and duration.  I'll change it to match your expectation.